text/0000-simple-postfix-macros.md

```diff
@@ -0,0 +1,171 @@
+- Feature Name: `simple_postfix_macros`
+- Start Date: 2018-05-12
+- RFC PR: (leave this empty)
+- Rust Issue: (leave this empty)
+
+# Summary
+[summary]: #summary
+
+Allow simple postfix macros, of the form `expr.ident!()`, to make macro
+invocations more readable and maintainable in left-to-right method chains.
+
+# Motivation
+[motivation]: #motivation
+
+The transition from `try!` to `?` doesn't just make error handling more
+concise; it allows reading expressions from left to right. An expression like
+`try!(try!(try!(foo()).bar()).baz())` required the reader to go back and forth
+between the left and right sides of the expression, and carefully match
+parentheses. The equivalent `foo()?.bar()?.baz()?` allows reading from left to
+right.
+
+The introduction of `await!` in RFC 2394 brings back the same issue, in the
+form of expressions like `await!(await!(await!(foo()).bar()).baz())`. This RFC
+would allow creating a postfix form of any such macro, simplifying that
+expression into a more readable `foo().await!().bar().await!().baz().await!()`.
+
+Previous discussions of method-like macros have stalled in the process of
+attempting to combine properties of macros (such as unevaluated arguments) with
+properties of methods (such as type-based or trait-based dispatch). This RFC
+proposes a minimal change to the macro system that allows defining a simple
+style of postfix macro, designed specifically for `await!` and for future cases
+like `try!` and `await!`, without blocking potential future extensions.
+
+# Guide-level explanation
+[guide-level-explanation]: #guide-level-explanation
+
+When defining a macro using `macro_rules!`, you can include a first argument
+that uses a designator of `self` (typically `$self:self`). This must appear as
+the first argument, and outside any repetition. If the macro includes such a
+case, then Rust code may invoke that macro using the method-like syntax
+`expr.macro!(args)`. The Rust compiler will expand the macro into code that
+receives the evaluated `expr` as its first argument.
+
+```rust
+macro_rules! log_value {
+    ($self:self, $msg:expr) => ({
```

Comment:
Just adding a new matcher `self` would probably be enough - `$anything: self`.
It wouldn't match anything except for a "method receiver" in `expr.my_macro!(...)`.

Reply:
🚲 🏠: I think in order to be consistent with function syntax it should be `$self` as in `($self, $msg:expr) => ({`.

Reply:
@petrochenkov That's exactly what I intended. I described the `:self` as a new "designator", because the Rust documentation used that term. Do you mean something different when you describe it as a "matcher"?

@est31 I considered that possibility; however, in addition to the inconsistency of not using a descriptor, that would limit potential future expansion a bit. Today, you can write `$self:expr` and use `$self`, without the compiler attaching any special meaning to the use of the name `self` as a macro argument. So, making `$self` without a descriptor special seems inconsistent in a problematic way.

Reply:
@joshtriplett 
I see, the RFC never shows an example with `$not_self: self`, so I thought that `$self` is mandatory.

Reply:
@petrochenkov Fixed.

Reply:
@joshtriplett good point. Ultimately I don't care much about the actual syntax.